File: python/dp/UniquePaths.py
'''
There is a robot on an m x n grid. The robot is initially located at the top-left corner (i.e., grid[0][0]). The robot tries to move to the bottom-right corner (i.e., grid[m - 1][n - 1]). The robot can only move either down or right at any point in time.

Given the two integers m and n, return the number of possible unique paths that the robot can take to reach the bottom-right corner.

Input: m = 3, n = 7
Output: 28

Input: m = 3, n = 2
Output: 3
Explanation: From the top-left corner, there are a total of 3 ways to reach the bottom-right corner:
1. Right -> Down -> Down
2. Down -> Down -> Right
3. Down -> Right -> Down
'''

class UniquePaths:
    def uniquePaths(self, m: int, n: int) -> int:
        # Initialize the bottom row with 1s
        # This represents that there is only 1 way to reach the end from any cell in the last row (by going right)
        row = [1] * n
        
        # Iterate from the second to last row up to the top row
        # We run this m-1 times because the last row is already initialized
        for i in range(m - 1):
            newRow = [1] * n
            # Iterate from right to left, starting from the second to last column
            # The last column is always 1 (can only go down)
            for j in range(n - 2, -1, -1):
                # The number of ways to reach the end from (i, j) is:
                # Ways from the right cell (newRow[j+1]) + Ways from the cell below (row[j])
                newRow[j] = newRow[j+1] + row[j]
            row = newRow
            
        return row[0]
        # Time Complexity: O(m * n)
        # Space Complexity: O(n) - we only store one row

if __name__ == "__main__":
    solution = UniquePaths()
    
    # Test Case 1
    m1, n1 = 3, 7
    print(f"Unique Paths for {m1}x{n1}: {solution.uniquePaths(m1, n1)}") # Expected: 28

    # Test Case 2
    m2, n2 = 3, 2
    print(f"Unique Paths for {m2}x{n2}: {solution.uniquePaths(m2, n2)}") # Expected: 3
    
    # Test Case 3
    m3, n3 = 7, 3
    print(f"Unique Paths for {m3}x{n3}: {solution.uniquePaths(m3, n3)}") # Expected: 28
    
    # Test Case 4
    m4, n4 = 3, 3
    print(f"Unique Paths for {m4}x{n4}: {solution.uniquePaths(m4, n4)}") # Expected: 6

File: src/main/java/com/naresh/a_dsalgo/dp/dp2/UniquePaths.java
package com.naresh.a_dsalgo.dp.dp2;

import java.util.Arrays;

/**
 * Problem: Unique Paths
 * Description: Find the number of possible unique paths from the top-left corner to the bottom-right corner of an m x n grid.
 * You can only move either down or right at any point in time.
 */
public class UniquePaths {
    /**
     * Algorithm: Bottom-up DP using a 2D array.
     * The number of paths to cell (i, j) is the sum of paths to (i-1, j) and (i, j-1).
     * Time: O(m * n), Space: O(m * n)
     */
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];

        // Initialize the first row and first column with 1s,
        // as there's only one way to reach any cell in them (by moving only right or only down).
        for (int[] row : dp) {
            Arrays.fill(row, 1);
        }

        // Fill the rest of the DP table
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }

        return dp[m - 1][n - 1];
    }

    /**
     * Algorithm: Bottom-up DP with space optimization.
     * The number of paths to cell (i, j) is the sum of paths to (i-1, j) and (i, j-1).
     * This version optimizes space to O(n) by only keeping track of the previous row.
     * Time: O(m * n), Space: O(n)
     */
    public int uniquePaths_SpaceOptimized(int m, int n) {
        // Pattern: DP (2D Space Optimized) | Time: O(m * n), Space: O(n)
        var row = new int[n];
        Arrays.fill(row, 1); // Initialize the first row (or current row acting as previous) with 1s
        for (int i = 1; i < m; i++) { // Iterate through rows starting from the second row
            for (int j = 1; j < n; j++) { // Iterate through columns starting from the second column
                row[j] += row[j - 1]; // Current cell's paths = paths from top (row[j]) + paths from left (row[j-1])
            }
        }
        return row[n - 1];
    }
    // FAANG Tip: Space can be optimized to O(min(m, n)). Mention the Combinatorics solution C(m+n-2, n-1) for O(1) space if applicable.

    public static void main(String[] args) {
        var m = 3;
        var n = 7;
        var solution = new UniquePaths();

        var result1 = solution.uniquePaths(m, n);
        System.out.println("Unique paths (O(m*n) DP) in " + m + "x" + n + " grid: " + result1); // Output: 28

        var result2 = solution.uniquePaths_SpaceOptimized(m, n);
        System.out.println("Unique paths (O(n) Space Optimized DP) in " + m + "x" + n + " grid: " + result2); // Output: 28
    }

    /*
     * Dry Run for uniquePaths_SpaceOptimized:
     * Input: m=3, n=7
     * Initialize row: [1,1,1,1,1,1,1]
     * After i=1 (second row calculation): [1,2,3,4,5,6,7]
     * After i=2 (third row calculation): [1,3,6,10,15,21,28]
     * Result: row[6]=28
     */
}
